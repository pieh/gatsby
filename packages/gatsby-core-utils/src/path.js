const path = require(`path`)
const os = require(`os`)

/**
 * @type {import('../index').joinPath}
 */
export function joinPath(...paths) {
  const joinedPath = path.join(...paths)
  if (os.platform() === `win32`) {
    return joinedPath.replace(/\\/g, `\\\\`)
  }

  return joinedPath
}
