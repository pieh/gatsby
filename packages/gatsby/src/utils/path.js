const { joinPath } = require(`gatsby-core-utils`)

export function withBasePath(basePath) {
  return (...paths) => joinPath(basePath, ...paths)
}

export function withTrailingSlash(basePath) {
  return `${basePath}/`
}
