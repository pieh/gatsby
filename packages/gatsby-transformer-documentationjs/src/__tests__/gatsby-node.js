import groupBy from "lodash/groupBy"
import path from "path"
import gatsbyNode from "../gatsby-node"

describe(`transformer-react-doc-gen: onCreateNode`, () => {
  let actions, node, createdNodes, updatedNodes
  const createNodeId = jest.fn(id => id)
  const createContentDigest = jest.fn().mockReturnValue(`content-digest`)

  let run = (node = node, opts = {}) =>
    gatsbyNode.onCreateNode(
      {
        node,
        actions,
        createNodeId,
        createContentDigest,
      },
      opts
    )

  beforeEach(() => {
    createdNodes = []
    updatedNodes = []
    node = {
      id: `node_1`,
      children: [],
      absolutePath: path.join(__dirname, `fixtures`, `code.js`),
      internal: {
        mediaType: `application/javascript`,
        type: `File`,
      },
    }
    actions = {
      createNode: jest.fn(n => createdNodes.push(n)),
      createParentChildLink: jest.fn(n => updatedNodes.push(n)),
    }
  })

  describe(`Simple example`, () => {
    it(`creates doc json apple node`, async () => {
      await run(node)

      const appleNode = createdNodes.find(node => node.name === `apple`)
      expect(appleNode).toBeDefined()
    })

    it(`should extract out a description, params, and examples`, async () => {
      await run(node)

      const appleNode = createdNodes.find(node => node.name === `apple`)

      expect(appleNode.examples.length).toBe(1)
      expect(appleNode.examples[0]).toMatchSnapshot(`example`)

      const appleDescriptionNode = createdNodes.find(
        node => node.id === appleNode.description___NODE
      )

      expect(appleDescriptionNode).toBeDefined()
      expect(appleDescriptionNode.internal.content).toMatchSnapshot(
        `description content`
      )

      const paramNode = createdNodes.find(
        node => node.id === appleNode.params___NODE[0]
      )

      expect(paramNode).toBeDefined()
      expect(paramNode.name).toMatchSnapshot(`param name`)

      const paramDescriptionNode = createdNodes.find(
        node => node.id === paramNode.description___NODE
      )

      expect(paramDescriptionNode).toBeDefined()
      expect(paramDescriptionNode.internal.content).toMatchSnapshot(
        `param description`
      )
    })
  })

  describe.skip(`Complex example`, () => {
    // TO-DO add tests
  })

  describe(`Sanity checks`, () => {
    it(`should only process javascript File nodes`, async () => {
      await run({ internal: { mediaType: `text/x-foo` } })
      expect(createdNodes.length).toBe(0)

      await run({ internal: { mediaType: `application/javascript` } })
      expect(createdNodes.length).toBe(0)

      await run({
        id: `test`,
        children: [],
        absolutePath: path.join(__dirname, `fixtures`, `code.js`),
        internal: { mediaType: `application/javascript`, type: `File` },
      })
      expect(createdNodes.length).toBeGreaterThan(0)
    })

    it(`should extract create description nodes with markdown types`, async () => {
      await run(node)
      let types = groupBy(createdNodes, `internal.type`)
      expect(
        types.DocumentationJSComponentDescription.every(
          d => d.internal.mediaType === `text/markdown`
        )
      ).toBe(true)
    })
  })
})
