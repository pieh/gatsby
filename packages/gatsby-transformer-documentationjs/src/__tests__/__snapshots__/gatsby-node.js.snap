// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformer-react-doc-gen: onCreateNode Simple example should extract out a description, params, and examples: description content 1`] = `
"A pretty cool jsdoc example
"
`;

exports[`transformer-react-doc-gen: onCreateNode Simple example should extract out a description, params, and examples: example 1`] = `
Object {
  "highlighted": "<span class=\\"token keyword\\">const</span> apple <span class=\\"token operator\\">=</span> <span class=\\"token function\\">require</span><span class=\\"token punctuation\\">(</span><span class=\\"token string\\">'apple'</span><span class=\\"token punctuation\\">)</span>
<span class=\\"token function\\">apple</span><span class=\\"token punctuation\\">(</span><span class=\\"token punctuation\\">)</span>",
  "raw": "const apple = require('apple')
apple()",
}
`;

exports[`transformer-react-doc-gen: onCreateNode Simple example should extract out a description, params, and examples: param description 1`] = `
"A nice crispy apple
"
`;

exports[`transformer-react-doc-gen: onCreateNode Simple example should extract out a description, params, and examples: param name 1`] = `"paramName"`;
